File: tests/java/TestHarness.java
import java.util.List;

import junit.framework.Assert;

import org.junit.Before;
import org.junit.Test;

import at.buy.api.APIException;
import at.buy.api.BuyatAPIClient;
import at.buy.api.BuyatAPIClient.FeedFormat;
import at.buy.api.BuyatAPIClient.ProductField;
import at.buy.api.BuyatAPIClient.SortOrder;
import at.buy.entities.Category;
import at.buy.entities.Feed;
import at.buy.entities.Product;
import at.buy.entities.ProductResultSet;
import at.buy.entities.Programme;

public class TestHarness {

  private BuyatAPIClient client;

  @Before
  public void setUp() {
    this.client = new BuyatAPIClient("01-fd288e15a739c35276a15d73a227dcc7");
  }

  @Test
  public void testProductSearch() {
    try {
      ProductResultSet result = client
          .searchProducts("XKJERVdsfkdjsfkDJHFKSJDFSDFDFkjhkwjean3endwf");
      Assert.assertEquals("Expected 0 results for nonsense query", 0, result
          .getTotalResults());
      Assert.assertEquals("Expected 0 results for nonsense query", 0, result
          .getProducts().size());

      result = client.searchProducts("ipod");
      Assert.assertNotNull("Result should not be null", result);
      Assert.assertTrue("Expected more than 10 results for ipod query", result
          .getTotalResults() > 10);
      Assert.assertEquals("Expected 10 results per page", result
          .getCurrentResults(), 10);

      List<Product> products = result.getProducts();
      Assert.assertNotNull("Products should not be null", products);
      Assert.assertEquals(
          "Products list should be same size as current results", result
              .getCurrentResults(), products.size());

      Product product = products.get(0);
      Assert.assertNotNull("Product should not be null", product);
      Assert.assertTrue("Product description should contain ipod", product
          .getDescription().toLowerCase().indexOf("ipod") != -1);

      Programme programme = product.getProgramme();
      Assert.assertNotNull("Programme should not be null", programme);
      Assert.assertEquals("Programme name should match product programme name",
          product.getProgrammeName(), programme.getProgrammeName());

      Feed feed = product.getFeed();
      Assert.assertNotNull("Feed should not be null", feed);
      Assert.assertEquals("Feed name should match product feed name",
          product.getFeedName(), feed.getFeedName());
      
      ProductResultSet result2 = client.searchProducts("Shoes", 2, 5, null, ProductField.ONLINE_PRICE, SortOrder.ASC);
      Assert.assertNotNull("Result should not be null", result2);
      Assert.assertTrue("Expected more than 10 results for shoes query", result2
          .getTotalResults() > 10);
      Assert.assertEquals("Expected 5 results per page", result2
          .getCurrentResults(), 5);
      Assert.assertTrue("1st product's price should be <= 2nd product's", 
          result2.getProducts().get(0).getOnlinePrice() <= result2.getProducts().get(1).getOnlinePrice());
      
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }
  
  @Test
  public void testCategories() {
    try {
      List<Category> level1Categories = client.getLevel1Categories();
      Assert.assertEquals("Number of categories", 38, level1Categories.size());
      
      Category cat = level1Categories.get(0);
      Assert.assertEquals("Category id", 20, cat.getCategoryID());
      Assert.assertEquals("Category name", "Baby", cat.getCategoryName());
      
      List<Category> subcategories = cat.getSubcategories();
      Assert.assertEquals("Number of categories", 9, subcategories.size());
      
      Category subcat = subcategories.get(0);
      Assert.assertEquals("Category id", 497, subcat.getCategoryID());
      Assert.assertEquals("Category name", "Accessories", subcat.getCategoryName());
      Assert.assertEquals("Parent category id", 20, subcat.getParentCategoryID());
      Assert.assertEquals("Parent category name", "Baby", subcat.getParentCategoryName());
      Assert.assertEquals("Subcategories", null, subcat.getSubcategories());
      
      List<Category> level2Categories = client.getLevel2Categories(20);
      Assert.assertEquals("Number of categories", subcategories.size(), level2Categories.size());
      
      List<Category> categories = client.getCategoryTree();
      Assert.assertEquals("Category tree should be same size as level 1 categories", level1Categories.size(), categories.size());
      
      for(int i=0; i<categories.size(); i++) {
        Category listCat = level1Categories.get(i);
        Category treeCat = categories.get(i);
        Assert.assertEquals("Category ids should match", listCat.getCategoryID(), treeCat.getCategoryID());
        Assert.assertEquals("Category names should match", listCat.getCategoryName(), treeCat.getCategoryName());
        //Assert.assertEquals("Category subcat list lengths should match", listCat.getSubcategories().size(), treeCat.getSubcategories().size());
      }
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }
  
  @Test
  public void testFeeds() {
    try {
      List<Feed> feeds = client.getFeeds();
      Assert.assertTrue("Expected more than 100 results", feeds.size()>100);
      
      Feed feed = feeds.get(0);
      Assert.assertEquals("Feed name", "1800Pet Meds", feed.getFeedName());
      Assert.assertEquals("Programme id", 1038, feed.getProgrammeID());
      Assert.assertEquals("Programme name", "1800PetMeds", feed.getProgrammeName());
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }
  
  @Test
  public void testFeedURL() {
    try {
      String feedURL = client.getFeedURL(367, FeedFormat.CSV, 0, 10, -1, -1, -1, null, "mylid", false, false, false);
      Assert.assertEquals("Feed URL", "http://localhost/feeddownload/index.php/download?OEMAIL=affiliate2@gmail.com&PX=e99a18c428cb38d5f260853678922e03&DISPLAYFORMAT=CSV&PRODUCTDB_ID=367&START=0&PERPAGE=10&LID=mylid", feedURL);

      String[] progCategories = {null, null, "Baby"};
      String feedURL2 = client.getFeedURL(367, FeedFormat.XMLGZIP, 0, 10, 20, -1, -1, progCategories, null, true, true, true);
      Assert.assertEquals("Feed URL", "https://localhost/feeddownload/index.php/download?OEMAIL=affiliate2@gmail.com&PX=e99a18c428cb38d5f260853678922e03&DISPLAYFORMAT=XMLGZIP&PRODUCTDB_ID=367&REVERSEMAPXML=true&START=0&PERPAGE=10&MAPPED_CAT_ID1=20&LEVEL3=Baby", feedURL2);
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }

  @Test
  public void testDeeplinks() {
    try {
      String url = client.getDeepLink("http://www.play.com/Music/CD/4-/793109/Sea-Change-The-Choral-Music-Of/Product.html?ptsl=1&ob=Price&fb=0");
      Assert.assertEquals("Deeplink URL", "http://playcom.at/affiliate2?CTY=84&DURL=http://www.play.com/Music/CD/4-/793109/Sea-Change-The-Choral-Music-Of/Product.html?ptsl=1", url);
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }

  @Test
  public void testProgrammes() {
    try {
      List<Programme> programmes = client.getProgrammes();
      Assert.assertTrue("Expected more than 100 results", programmes.size()>100);
      
      Programme prog = programmes.get(0);
      Assert.assertEquals("Programme id", 115, prog.getProgrammeID());
      Assert.assertEquals("Programme name", "Marks and Spencer", prog.getProgrammeName());
      
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }
  
  @Test
  public void testProductInfo() {
    try {
      Product product = client.getProductInfo(118949289);
      Assert.assertNotNull("Product should not be null", product);
      Assert.assertEquals("Product id", 118949289, product.getProductID());
      Assert.assertEquals("Product SKU", "90998", product.getProductSKU());
      Assert.assertEquals("Product name", "St  James's Hotel and Club - Althoff  Hotel Collection", product.getProductName());
      Assert.assertEquals("Product url", "http://laterooms.at/affiliate2?CTY=84&DURL=http://www.laterooms.com/en/p1023/hotel-reservations/90998_st-james-s-club-hotel.aspx", product.getProductURL());
      Assert.assertEquals("Prog id", 293, product.getProgrammeID());
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }
  
  @Test
  public void testProgrammeInfo() {
    try {
      Programme programme = client.getProgrammeInfo(293);
      Assert.assertNotNull("Programme should not be null", programme);
      Assert.assertEquals("Programme id", 293, programme.getProgrammeID());
      Assert.assertEquals("Programme name", "Laterooms.com", programme.getProgrammeName());
      Assert.assertEquals("Programme url", "http://laterooms.at/affiliate2", programme.getProgrammeURL());
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }
  
  @Test
  public void testFeedInfo() {
    try {
      Feed feed = client.getFeedInfo(367);
      Assert.assertNotNull("Feed should not be null", feed);
      Assert.assertEquals("Feed id", 367, feed.getFeedID());
      Assert.assertEquals("Feed name", "1800Pet Meds", feed.getFeedName());
      Assert.assertEquals("Programme id", 1038, feed.getProgrammeID());
      Assert.assertEquals("Programme name", "1800PetMeds", feed.getProgrammeName());
    } catch (APIException e) {
      e.printStackTrace();
      Assert.assertTrue("Got exception: " + e.getMessage(), false);
    }
  }
}

File: tests/python/TestHarness.py
'''
Created on 3 Dec 2009

@author: aclarke
'''
import unittest
from buyat.api.BuyatAPIClient import BuyatAPIClient

class TestHarness(unittest.TestCase):
    
    def setUp(self):
        self.client = BuyatAPIClient('01-fd288e15a739c35276a15d73a227dcc7');
        
        
    def testProductSearch(self):
        result = self.client.search_products('XKJERVdsfkdjsfkDJHFKSJDFSDFDFkjhkwjean3endwf')
        assert result.total_results == 0, result.total_results
        assert len(result.products) == 0, len(result.products)
        
        result = self.client.search_products('ipod')
        assert result.total_results > 10, result.total_results
        assert result.current_results == 10, result.current_results
        assert result.current_results == len(result.products), (result.total_results, len(result.products))
        
        assert len(result.products[0].description.lower()) > 0, len(result.products[0].description.lower())
        
        
    def testCategories(self):
        categories = self.client.get_level1_categories()
        assert len(categories) == 38, len(categories)
        
        assert categories[0].category_id == 20, categories[0].category_id
        assert categories[0].category_name == 'Baby', categories[0].category_name
        assert len(categories[0].subcategories) == 9, len(categories[0].subcategories)
            
        subcategories = self.client.get_level2_categories(categories[0].category_id)
        assert len(subcategories) == 9, len(subcategories)
        assert subcategories[0].subcategories == None, subcategories[0].subcategories
            
        both_subcats = zip(subcategories, categories[0].subcategories)
        for cat1, cat2 in both_subcats:
            assert cat1.category_id == cat2.category_id
            assert cat1.category_name == cat2.category_name
            assert cat1.level == cat2.level
            
        category_tree = self.client.get_category_tree()
        assert len(category_tree) == 38, len(category_tree)
        
        both_cats = zip(categories, category_tree)
        for cat1, cat2 in both_cats:
            assert cat1.category_id == cat2.category_id
            assert cat1.category_name == cat2.category_name
        
        
    def testProgrammes(self):
        programmes = self.client.get_programmes()
        assert len(programmes) == 273, len(programmes)
        
        assert programmes[0].programme_id == 115, programmes[0].programme_id
        assert programmes[0].programme_name == "Marks and Spencer", programmes[0].programme_name
        
        programme = self.client.get_programme(115)
        assert programme.programme_id == 115, programme.programme_id
        assert programme.programme_name == "Marks and Spencer", programme.programme_name
        assert programme.programme_url == programmes[0].programme_url, (programme.programme_url, programmes[0].programme_url)
        assert programme.has_feed == programmes[0].has_feed, (programme.has_feed, programmes[0].has_feed)
        
        
    def testFeeds(self):
        feeds = self.client.get_feeds()
        assert len(feeds) > 100, len(feeds)
        
        assert feeds[0].feed_id == 367, feeds[0].feed_id
        assert feeds[0].feed_name == "1800Pet Meds", feeds[0].feed_name
        
        feed = self.client.get_feed(367)
        assert feed.feed_id == 367, feed.feed_id
        assert feed.feed_name == "1800Pet Meds", feed.feed_name
        assert feed.programme_name == feeds[0].programme_name, (feed.programme_name, feeds[0].programme_name)
        assert feed.programme_url == feeds[0].programme_url, (feed.programme_url, feeds[0].programme_url)
        assert feed.number_of_products == feeds[0].number_of_products, (feed.number_of_products, feeds[0].number_of_products)
        assert feed.last_updated == feeds[0].last_updated, (feed.last_updated, feeds[0].last_updated)
        
    def testFeedURL(self):
        url = self.client.get_feed_url(367, "CSV", perpage=10, lid="mylid")
        assert url == "http://localhost/feeddownload/index.php/download?OEMAIL=affiliate2@gmail.com&amp;PX=e99a18c428cb38d5f260853678922e03&amp;DISPLAYFORMAT=CSV&amp;PRODUCTDB_ID=367&amp;START=0&amp;PERPAGE=10&amp;LID=mylid", url
        
        prog_categories = [None, None, "Baby"]
        url2 = self.client.get_feed_url(367, "XMLGZIP", perpage=10, level1_category_id=20, levels=prog_categories, reverse_map_xml=True, use_https=True);
        assert url2 == "https://localhost/feeddownload/index.php/download?OEMAIL=affiliate2@gmail.com&amp;PX=e99a18c428cb38d5f260853678922e03&amp;DISPLAYFORMAT=XMLGZIP&amp;PRODUCTDB_ID=367&amp;REVERSEMAPXML=y&amp;START=0&amp;PERPAGE=10&amp;MAPPED_CAT_ID1=20&amp;LEVEL3=Baby", url2
        
        
    def testGetDeeplink(self):
        url = self.client.get_deeplink("http://www.play.com/Music/CD/4-/793109/Sea-Change-The-Choral-Music-Of/Product.html?ptsl=1&ob=Price&fb=0");
        assert url == "http://playcom.at/affiliate2?CTY=84&amp;DURL=http://www.play.com/Music/CD/4-/793109/Sea-Change-The-Choral-Music-Of/Product.html?ptsl=1", url

    def testProductInfo(self):
        product = self.client.get_product(118949289)
        assert product.product_id == 118949289
        assert product.product_sku == "90998"
        assert product.product_name == "St  James's Hotel and Club - Althoff  Hotel Collection"
        assert product.product_url == "http://laterooms.at/affiliate2?CTY=84&amp;DURL=http://www.laterooms.com/en/p1023/hotel-reservations/90998_st-james-s-club-hotel.aspx"
        
        feed = product.feed
        assert feed.feed_name == product.feed_name
        assert feed.feed_id == product.feed_id
        
        programme = product.programme
        assert programme.programme_name == product.programme_name
        assert programme.programme_id == product.programme_id
        assert programme.programme_url == product.programme_url, (programme.programme_url, product.programme_url)
        
        

if __name__ == "__main__":
    #import sys;sys.argv = ['', 'Test.testName']
    unittest.main()